misc/metrics/src/bandwidth.rs

```diff
@@ -0,0 +1,312 @@
+use crate::protocol_stack;
+use futures::{
+    future::{MapOk, TryFutureExt},
+    io::{IoSlice, IoSliceMut},
+    prelude::*,
+    ready,
+};
+use libp2p_core::{
+    muxing::{StreamMuxer, StreamMuxerEvent},
+    transport::{ListenerId, TransportError, TransportEvent},
+    Multiaddr,
+};
+use libp2p_identity::PeerId;
+use prometheus_client::{
+    encoding::{EncodeLabelSet, EncodeLabelValue},
+    metrics::{counter::Counter, family::Family},
+    registry::{Registry, Unit},
+};
+use std::{
+    convert::TryFrom as _,
+    io,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+#[derive(Debug, Clone)]
+#[pin_project::pin_project]
+pub struct Transport<T> {
+    #[pin]
+    transport: T,
+    metrics: Family<Labels, Counter>,
+}
+
+impl<T> Transport<T> {
+    pub fn new(transport: T, registry: &mut Registry) -> Self {
+        let metrics = Family::<Labels, Counter>::default();
+        registry
+            .sub_registry_with_prefix("libp2p")
+            .register_with_unit(
+                "bandwidth",
+                "Bandwidth usage by direction and transport protocols",
+                Unit::Bytes,
+                metrics.clone(),
+            );
```

Comment:
Is this appending the `unit` automatically? i.e. will this result in a `bandwidth_bytes` metric?

Reply:
Correct. It does all the appending and additional `# UNIT` line that is required by the Open Metrics specification.

Reply:
I think in several (documentation) places, you mention it without the unit though. Might be worth fixing.

Reply:
Should be fixed with https://github.com/libp2p/rust-libp2p/pull/4727/commits/bd83ba978cc368e01b2066cc0a96f30e9021483b.